// pnnl-react-core/lib/Sidebar/Sidebar.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Col = require('react-bootstrap/lib/Col');

var _Col2 = _interopRequireDefault(_Col);

var _elementType = require('prop-types-extra/lib/elementType');

var _elementType2 = _interopRequireDefault(_elementType);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Sidebar = function Sidebar(_ref) {
  var componentClass = _ref.componentClass,
      xs = _ref.xs,
      sm = _ref.sm,
      md = _ref.md,
      lg = _ref.lg,
      fixed = _ref.fixed,
      children = _ref.children;

  var Component = componentClass || _Col2['default'];
  var itemClasses = (0, _classnames2['default'])('PnnlSidebar', { fixed: fixed });

  return _react2['default'].createElement(
    Component,
    { className: itemClasses, xs: xs, sm: sm, md: md, lg: lg },
    children
  );
};

Sidebar.propTypes = {
  /**
   * The name of the element or component used to build the base Sidebar. Defaults to Col.
   */
  componentClass: _elementType2['default'],
  /**
   * The number of grid columns the Sidebar should take up at the extra-small (xs) breakpoint.
   */
  xs: _propTypes2['default'].number,
  /**
   * The number of grid columns the Sidebar should take up at the small (sm) breakpoint.
   */
  sm: _propTypes2['default'].number,
  /**
   * The number of grid columns the Sidebar should take up at the medium (md) breakpoint.
   */
  md: _propTypes2['default'].number,
  /**
   * The number of grid columns the Sidebar should take up at the large (lg) breakpoint.
   */
  lg: _propTypes2['default'].number,
  /**
   * Whether the Sidebar is fixed, or whether it scrolls with the page.
   */
  fixed: _propTypes2['default'].bool,
  /**
   * The content to be displayed in the Sidebar.  Often a combination of SidebarPanel, SidebarItem, Timeline, or other form inputs.
   * @required
   */
  children: _propTypes2['default'].node.isRequired
};

Sidebar.defaultProps = {
  componentClass: null,
  xs: 12,
  sm: 4,
  md: 3,
  lg: 3,
  fixed: false
};

exports['default'] = Sidebar;
module.exports = exports['default'];
