// pnnl-react-core/lib/NavMenu/NavMenu.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Dropdown = require('react-bootstrap/lib/Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

var _Action = require('../Action');

var _Action2 = _interopRequireDefault(_Action);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/*
<PnnlNavMenu icon="thumbs-up" text="Feedback" id="feedbackMenu">
    <LinkContainer to="/feedback">
        <MenuItem>Positive</MenuItem>
    </LinkContainer>
    <LinkContainer to="/feedback">
        <MenuItem>Negative</MenuItem>
    </LinkContainer>
</PnnlNavMenu>

NOTE: LinkContainer comes from react-router-bootstrap, MenuItem from react-bootstrap.
You only need LinkContainer if you want to have the menu items be links without attaching
any actions to them yourself, otherwise MenuItem will work just fine.
*/

var NavMenu = function NavMenu(_ref) {
  var id = _ref.id,
      icon = _ref.icon,
      text = _ref.text,
      children = _ref.children;

  var restProps = {
    icon: icon,
    text: text,
    isMenu: true,
    iconDisplay: 'all'
  };

  return _react2['default'].createElement(
    _Dropdown2['default'],
    { id: id, componentClass: 'li', className: 'PnnlNavMenu' },
    _react2['default'].createElement(_Action2['default'], (0, _extends3['default'])({ bsRole: 'toggle' }, restProps)),
    _react2['default'].createElement(
      _Dropdown2['default'].Menu,
      { bsRole: 'menu' },
      children
    )
  );
};

NavMenu.propTypes = {
  /**
   * The name of the icon to be displayed in the NavItem.
   * @required
   */
  icon: _propTypes2['default'].string.isRequired,

  /**
   * The label to be displayed underneath the icon in the NavItem.
   * @required
   */
  text: _propTypes2['default'].string.isRequired,

  /**
   * The items to be displayed in a Menu.  These will likely be LinkContainer components wrapping MenuItem components.
   * @required
   */
  children: _propTypes2['default'].oneOfType([_propTypes2['default'].arrayOf(_propTypes2['default'].node), _propTypes2['default'].node]).isRequired,

  /**
   * A unique ID so that react-bootstrap can link the button with the menu it toggles.
   * @required
   */
  id: _propTypes2['default'].oneOfType([_propTypes2['default'].string, _propTypes2['default'].number]).isRequired
};

exports['default'] = NavMenu;
module.exports = exports['default'];
