// pnnl-react-core/lib/config.js
'use strict';

exports.__esModule = true;
exports['default'] = {
  defaultPersonImage: 'https://cdn.pnl.gov/images/defaults/person.png',
  defaultFacilityImage: 'https://cdn.pnl.gov/images/defaults/facility.png',
  defaultLocationImage: 'https://cdn.pnl.gov/images/defaults/space.png',
  defaultSpaceImage: 'https://cdn.pnl.gov/images/defaults/space.png',
  scout: 'https://scout.pnl.gov/api',
  mitLocation: function mitLocation(locationId) {
    return locationId ? 'https://mit2.pnl.gov/?v=map&locid=' + locationId : '';
  },
  deptLocation: function deptLocation(deptId) {
    return deptId ? 'https://hr.pnl.gov/hrdrilldown/default.asp?deptid=' + deptId : '';
  },
  opwhse: {
    self: 'https://api.pnl.gov/operations/v2/people/self',
    selfImage: 'https://api.pnl.gov/operations/v2/portraits/self?type=image',
    person: function person(hid) {
      return hid ? 'https://api.pnl.gov/operations/v2/people/' + hid + '/?expand=space,facility,department' : '';
    },
    facility: function facility(id) {
      return id ? 'https://api.pnl.gov/operations/v2/facilities/' + id : '';
    },
    location: function location(id) {
      return id ? 'https://api.pnl.gov/operations/v2/locations/' + id + '?expand=facility,space' : '';
    },
    space: function space(id) {
      return id ? 'https://api.pnl.gov/operations/v2/spaces/' + id + '?expand=facility' : '';
    },
    personImage: function personImage(hid) {
      return hid ? 'https://api.pnl.gov/operations/v2/portraits/' + hid + '?type=image' : '';
    },
    locationImage: function locationImage(locationId) {
      return locationId ? 'https://api.pnl.gov/operations/v2/locations/' + locationId + '/image' : '';
    }
  },
  themeCore: {
    svgIcons: 'https://api.pnl.gov/it/theme-core/v1/svg-icons'
  }
};
module.exports = exports['default'];
