// pnnl-react-core/lib/CardMedia/CardMedia.js
'use strict';

exports.__esModule = true;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var CardMedia = function CardMedia(_ref) {
  var backgroundColorClass = _ref.backgroundColorClass,
      imageUrl = _ref.imageUrl,
      className = _ref.className,
      children = _ref.children,
      tint = _ref.tint,
      style = _ref.style;

  var mediaStyle = {
    backgroundImage: 'url(' + imageUrl
  };
  var _style = (0, _assign2['default'])({}, mediaStyle, style);

  var mediaClasses = (0, _classnames2['default'])('PnnlCard-media', backgroundColorClass, className);

  var tintClasses = (0, _classnames2['default'])('PnnlCard-mediaTint', tint);

  return _react2['default'].createElement(
    'div',
    { className: mediaClasses, style: _style },
    _react2['default'].createElement(
      'div',
      { className: tintClasses },
      children
    )
  );
};

CardMedia.propTypes = {
  /**
   * Background color to be used if there's no imageUrl supplied. Must be a class that corresponds to one of our brand colors.
   */
  backgroundColorClass: _propTypes2['default'].string,
  /**
    * Image to be displayed.
    * */
  imageUrl: _propTypes2['default'].string,
  /**
   * Content to be displayed within the media block.
   */
  children: _propTypes2['default'].node,
  /**
   * A color to overlay atop an image.  Will also produce a blur on the image for browsers that support it.
   */
  tint: _propTypes2['default'].string
};

CardMedia.defaultProps = {
  backgroundColorClass: 'copper',
  imageUrl: null,
  children: null,
  tint: null
};

exports['default'] = CardMedia;
module.exports = exports['default'];
