// pnnl-react-core/src/Sidebar/Sidebar.js
import React from 'react';
import PropTypes from 'prop-types';
import Col from 'react-bootstrap/lib/Col';
import elementType from 'prop-types-extra/lib/elementType';
import classNames from 'classnames';

const Sidebar = ({
  componentClass, xs, sm, md, lg, fixed, children,
}) => {
  const Component = (componentClass) || Col;
  const itemClasses = classNames('PnnlSidebar', { fixed });

  return (
    <Component className={itemClasses} xs={xs} sm={sm} md={md} lg={lg}>
      {children}
    </Component>
  );
};

Sidebar.propTypes = {
  /**
   * The name of the element or component used to build the base Sidebar. Defaults to Col.
   */
  componentClass: elementType,
  /**
   * The number of grid columns the Sidebar should take up at the extra-small (xs) breakpoint.
   */
  xs: PropTypes.number,
  /**
   * The number of grid columns the Sidebar should take up at the small (sm) breakpoint.
   */
  sm: PropTypes.number,
  /**
   * The number of grid columns the Sidebar should take up at the medium (md) breakpoint.
   */
  md: PropTypes.number,
  /**
   * The number of grid columns the Sidebar should take up at the large (lg) breakpoint.
   */
  lg: PropTypes.number,
  /**
   * Whether the Sidebar is fixed, or whether it scrolls with the page.
   */
  fixed: PropTypes.bool,
  /**
   * The content to be displayed in the Sidebar.  Often a combination of SidebarPanel, SidebarItem, Timeline, or other form inputs.
   * @required
   */
  children: PropTypes.node.isRequired,
};

Sidebar.defaultProps = {
  componentClass: null,
  xs: 12,
  sm: 4,
  md: 3,
  lg: 3,
  fixed: false,
};

export default Sidebar;
