// pnnl-react-core/src/CardMedia/CardMedia.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const CardMedia = ({
  backgroundColorClass, imageUrl, className, children, tint, style
}) => {
  const mediaStyle = {
    backgroundImage: `url(${imageUrl}`
  };
  const _style = Object.assign({}, mediaStyle, style);

  const mediaClasses = classNames(
    'PnnlCard-media',
    backgroundColorClass,
    className
  );

  const tintClasses = classNames(
    'PnnlCard-mediaTint',
    tint
  );

  return (
    <div className={mediaClasses} style={_style}>
      <div className={tintClasses}>
        {children}
      </div>
    </div>
  );
};

CardMedia.propTypes = {
  /**
   * Background color to be used if there's no imageUrl supplied. Must be a class that corresponds to one of our brand colors.
   */
  backgroundColorClass: PropTypes.string,
  /**
    * Image to be displayed.
    * */
  imageUrl: PropTypes.string,
  /**
   * Content to be displayed within the media block.
   */
  children: PropTypes.node,
  /**
   * A color to overlay atop an image.  Will also produce a blur on the image for browsers that support it.
   */
  tint: PropTypes.string
};

CardMedia.defaultProps = {
  backgroundColorClass: 'copper',
  imageUrl: null,
  children: null,
  tint: null,
};

export default CardMedia;
