// pnnl-react-core/src/SidebarHeader/SidebarHeader.js
import React from 'react';
import PropTypes from 'prop-types';
import Panel from 'react-bootstrap/lib/Panel';
import Icon from '../Icon';
import ValidationIcon from '../ValidationIcon';

class SidebarHeader extends React.Component {
  constructor(props) {
    super(props);

    this.renderValidation = this.renderValidation.bind(this);
  }

  renderValidation() {
    const { validationState } = this.props;

    return (<ValidationIcon validationState={validationState} />);
  }

  render() {
    const {
      children, collapsible, validationState, title, open
    } = this.props;
    const iconName = (open) ? 'chevron-up' : 'chevron-down';

    return (
      <Panel.Heading>
        <Panel.Title toggle={collapsible}>
          {validationState && this.renderValidation()}
          {title}
          {collapsible && <span className="arrow"><Icon icon={iconName} /></span>}
        </Panel.Title>
        {children}
      </Panel.Heading>
    );
  }
}

SidebarHeader.propTypes = {
  /**
   * Whether the SidebarHeader is part of a collapsible SidebarPanel.
   */
  collapsible: PropTypes.bool,
  /**
   * Whether the SidebarPanel connected to the SidebarHeader is open, or not.
   */
  open: PropTypes.bool,
  /**
   * The label to be displayed in the header.
   * @required
   */
  title: PropTypes.string.isRequired,
  /**
   * If the SidebarPanel connected to this SidebarHeader has validation, display an icon summarizing the validation state. Adds a span and Icon before the title.
   */
  validationState: PropTypes.oneOf(['success', 'error', 'incomplete', 'warning']),
  /**
   * Additional components to be added to the SidebarHeader.  Actions and Dropdowns are common items.
   */
  children: PropTypes.node,
};

SidebarHeader.defaultProps = {
  collapsible: false,
  open: true,
  validationState: null,
  children: null,
};

export default SidebarHeader;
