// pnnl-react-core/src/Location/Location.js
import React from 'react';
import PropTypes from 'prop-types';
import config from '../config';
import Facility from '../Facility';
import Space from '../Space';
import Loader from '../Loader';
import api from '../__helpers/restApiHelper';

export default class Location extends React.Component {
  static propTypes = {
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /**
      * A Facility object for the facility you want to display.
      *
      *
      * The object needs to have the same naming convention as: https://api.pnl.gov/operations/v2/locations/{id}?expand=facility
      * @required
      * */
    location: PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
      thumbnailImage: PropTypes.string,
      isFacility: PropTypes.bool,
      facility: PropTypes.shape({
        id: PropTypes.string,
        name: PropTypes.string,
      }),
    }),

    /**
     * Which side should the Avatar be aligned on? Choosing 'center' will stack the information
     */
    align: PropTypes.oneOf(['left', 'right', 'center']),

    /**
      * Callback you want to run if the Person component is `clicked`.
      * */
    onClick: PropTypes.func,

    /**
     * The size of the Avatar to be generated.
     */
    size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
  };

  static defaultProps = {
    align: 'left',
    onClick: undefined,
    size: 'md',
    location: undefined,
    id: undefined,
  };


  constructor(props) {
    super(props);
    this.state = {
      data: this.props.location || undefined,
    };
  }

  componentDidMount() {
    if (!this.props.id && !this.props.location) {
      throw new Error('A location identifier was not supplied. This can be either props.id (Location Id) or props.location.id.');
    }
    if (this.props.id && !this.props.location) {
      api.GET(config.opwhse.location(this.props.id), {
        signal: this.abortController && this.abortController.signal
      })
        .then((data) => {
          this.setState({
            data,
          });
        })
        .catch((error) => {
          throw new Error(`An error resulted when trying to get the location from the API. ${error}`);
        });
    }
  }

  componentWillUnmount() {
    if (this.abortController) {
      this.abortController.abort();
    }
  }

  abortController = (window && window.AbortController) ? new window.AbortController() : undefined;

  render() {
    const { data } = this.state;

    if (!data) {
      return (
        <Loader show />
      );
    }

    if (data.isFacility === true) {
      return (
        <Facility facility={data.facility} {...this.props} />
      );
    }
    if (data.isSpace === true) {
      return (
        <Space space={{ ...data.space, facility: data.facility }} {...this.props} />
      );
    }
    return (
      <div>Unknown Location Type</div>
    );
  }
}
