// pnnl-react-core/lib/Action/Action.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Button = require('react-bootstrap/lib/Button');

var _Button2 = _interopRequireDefault(_Button);

var _propTypesExtra = require('prop-types-extra');

var _Icon = require('../Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _Link = require('../Link');

var _Link2 = _interopRequireDefault(_Link);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Action = function (_React$Component) {
  (0, _inherits3['default'])(Action, _React$Component);

  function Action() {
    (0, _classCallCheck3['default'])(this, Action);
    return (0, _possibleConstructorReturn3['default'])(this, _React$Component.apply(this, arguments));
  }

  Action.prototype.renderLink = function renderLink(actionClasses) {
    var _props = this.props,
        icon = _props.icon,
        text = _props.text,
        to = _props.to;

    var restProps = {
      onlyActiveOnIndex: this.props.onlyActiveOnIndex,
      activeClassName: 'active'
    };
    return _react2['default'].createElement(
      _Link2['default'],
      (0, _extends3['default'])({ to: to, className: actionClasses }, restProps),
      _react2['default'].createElement(_Icon2['default'], (0, _extends3['default'])({ icon: icon }, this.props)),
      _react2['default'].createElement(
        'small',
        null,
        text
      )
    );
  };

  Action.prototype.render = function render() {
    var _props2 = this.props,
        icon = _props2.icon,
        text = _props2.text,
        to = _props2.to,
        displayType = _props2.displayType,
        bsRole = _props2.bsRole,
        componentClass = _props2.componentClass,
        id = _props2.id,
        onClick = _props2.onClick,
        onKeyDown = _props2.onKeyDown,
        readonly = _props2.readonly,
        children = _props2.children;

    var actionClasses = (0, _classnames2['default'])(displayType, bsRole, 'PnnlAction');

    if (to) {
      return this.renderLink(actionClasses);
    }

    var caretClassName = bsRole ? 'caret' : '';
    var Component = componentClass || _Button2['default'];

    return _react2['default'].createElement(
      Component,
      {
        id: id,
        className: actionClasses,
        bsRole: bsRole,
        onClick: onClick,
        onKeyDown: onKeyDown,
        disabled: readonly
      },
      _react2['default'].createElement(_Icon2['default'], (0, _extends3['default'])({ icon: icon }, this.props)),
      text !== null && _react2['default'].createElement(
        'small',
        null,
        text
      ),
      _react2['default'].createElement('span', { className: caretClassName }),
      children
    );
  };

  return Action;
}(_react2['default'].Component);

Action.propTypes = {
  icon: _propTypes2['default'].string.isRequired,
  text: _propTypes2['default'].string,
  /** if 'to' is specified, we'll render a <Link>, otherwise a <button> */
  to: _propTypes2['default'].oneOfType([_propTypes2['default'].string, _propTypes2['default'].object]),
  componentClass: _propTypesExtra.elementType,
  displayType: _propTypes2['default'].string,
  readonly: _propTypes2['default'].bool,
  /** only show link to '/' as active if actually on that page */
  onlyActiveOnIndex: _propTypes2['default'].bool,
  bsRole: _propTypes2['default'].string,

  // these are pass throughs and don't need defauls to
  /* eslint-disable react/require-default-props */
  id: _propTypes2['default'].oneOfType([_propTypes2['default'].string, _propTypes2['default'].number]),
  onClick: _propTypes2['default'].func,
  onKeyDown: _propTypes2['default'].func
  /* eslint-enable */
};

Action.defaultProps = {
  onlyActiveOnIndex: true,
  readonly: false,
  componentClass: null,
  displayType: null,
  bsRole: null,
  text: null,
  to: null
};

exports['default'] = Action;
module.exports = exports['default'];
