// pnnl-react-core/lib/PageHeader/PageHeader.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Grid = require('react-bootstrap/lib/Grid');

var _Grid2 = _interopRequireDefault(_Grid);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var PageHeader = function PageHeader(_ref) {
  var children = _ref.children,
      fluid = _ref.fluid,
      className = _ref.className,
      fixed = _ref.fixed;

  var headerClasses = (0, _classnames2['default'])(className, 'PnnlPageHeader', { fixed: fixed });

  return _react2['default'].createElement(
    'div',
    { className: headerClasses },
    _react2['default'].createElement(
      _Grid2['default'],
      { fluid: fluid },
      children
    )
  );
};

PageHeader.propTypes = {
  /**
     * The items to display in the PageHeader.  Should include one or more `<Row />` components.
     */
  children: _propTypes2['default'].oneOfType([_propTypes2['default'].arrayOf(_propTypes2['default'].node), _propTypes2['default'].node]).isRequired,
  /**
     * An additional CSS class to be added to the root PageHeader element.
     */
  className: _propTypes2['default'].string,
  /**
     * Determines whether the Container is fluid or fixed-width.
     */
  fluid: _propTypes2['default'].bool,
  /**
   * Determines whether the PageHeader is fixed, or scrolls with the page.
   */
  fixed: _propTypes2['default'].bool
};

PageHeader.defaultProps = {
  className: '',
  fluid: false,
  fixed: false
};

exports['default'] = PageHeader;
module.exports = exports['default'];
