// pnnl-react-core/src/config.js
export default {
  defaultPersonImage: 'https://cdn.pnl.gov/images/defaults/person.png',
  defaultFacilityImage: 'https://cdn.pnl.gov/images/defaults/facility.png',
  defaultLocationImage: 'https://cdn.pnl.gov/images/defaults/space.png',
  defaultSpaceImage: 'https://cdn.pnl.gov/images/defaults/space.png',
  scout: 'https://scout.pnl.gov/api',
  mitLocation: locationId => (locationId ? `https://mit2.pnl.gov/?v=map&locid=${locationId}` : ''),
  deptLocation: deptId => (deptId ? `https://hr.pnl.gov/hrdrilldown/default.asp?deptid=${deptId}` : ''),
  opwhse: {
    self: 'https://api.pnl.gov/operations/v2/people/self',
    selfImage: 'https://api.pnl.gov/operations/v2/portraits/self?type=image',
    person: hid => (hid ? `https://api.pnl.gov/operations/v2/people/${hid}/?expand=space,facility,department` : ''),
    facility: id => (id ? `https://api.pnl.gov/operations/v2/facilities/${id}` : ''),
    location: id => (id ? `https://api.pnl.gov/operations/v2/locations/${id}?expand=facility,space` : ''),
    space: id => (id ? `https://api.pnl.gov/operations/v2/spaces/${id}?expand=facility` : ''),
    personImage: hid => (hid ? `https://api.pnl.gov/operations/v2/portraits/${hid}?type=image` : ''),
    locationImage: locationId => (locationId ? `https://api.pnl.gov/operations/v2/locations/${locationId}/image` : ''),
  },
  themeCore: {
    svgIcons: 'https://api.pnl.gov/it/theme-core/v1/svg-icons',
  },
};
