// pnnl-react-core/src/Facility/Facility.js
import React from 'react';
import PropTypes from 'prop-types';
import config from '../config';
import Avatar from '../Avatar';
import AvatarInfo from '../AvatarInfo';
import Loader from '../Loader';
import api from '../__helpers/restApiHelper';

export default class Facility extends React.Component {
  static propTypes = {
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /**
      * A facility object for the facility you want display.
      *
      *
      * The object needs to have the same naming convention as: https://api.pnl.gov/operations/v2/facilities/{id}
      * @required
      * */
    facility: PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
      abbreviation: PropTypes.string,
      thumbnailImage: PropTypes.string,
      address: PropTypes.shape({ line1: PropTypes.string }),
    }),

    /**
     * Which side should the Avatar be aligned on? Choosing 'center' will stack the information
     */
    align: PropTypes.oneOf(['left', 'right', 'center']),

    /**
      * Callback you want to run if the Person component is `clicked`.
      * */
    onClick: PropTypes.func,

    /**
     * The size of the Avatar to be generated.
     */
    size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
  };

  static defaultProps = {
    align: 'left',
    onClick: undefined,
    size: 'md',
    facility: undefined,
    id: undefined,
  };

  constructor(props) {
    super(props);
    this.state = {
      data: this.props.facility || undefined,
    };
  }

  componentDidMount() {
    if (!this.props.id && !this.props.facility.id) {
      throw new Error('A facility identifier was not supplied. This can be either props.id (Facility Id) or props.facility.id.');
    }
    if (this.props.id && !this.props.facility) {
      api.GET(config.opwhse.facility(this.props.id), {
        signal: this.abortController && this.abortController.signal
      })
        .then(data =>
          this.setState({
            data,
          }))
        .catch((error) => {
          throw new Error(`An error resulted when trying to get the facility from the API. ${error}`);
        });
    }
  }

  componentWillUnmount() {
    if (this.abortController) {
      this.abortController.abort();
    }
  }

  abortController = (window && window.AbortController) ? new window.AbortController() : undefined;

  render() {
    const { data } = this.state;
    const { onClick, size, align } = this.props;

    if (!data) {
      return (
        <Loader show />
      );
    }

    const address = ((data.address || {}).line1 || '').replace(
      /\w\S*/g,
      txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()
    );
    const avatarComponent =
      (<Avatar src={data.thumbnailImage || config.defaultFacilityImage} size={size} />);

    const avatarInfo = (
      <AvatarInfo avatarComponent={avatarComponent} title={data.name} subTitle={address} align={align} />
    );

    if (onClick) {
      return (
        <button className="PnnlFacility" onClick={onClick}>
          {avatarInfo}
        </button>
      );
    }
    return (
      <div className="PnnlFacility">
        {avatarInfo}
      </div>
    );
  }
}
