// react-scripts-pnnl-boilerplate/template-typescript/src/pages/Home.jsx
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { LinkContainer } from 'react-router-bootstrap';
import Grid from 'react-bootstrap/es/Grid';
import Button from 'react-bootstrap/es/Button';
import ProfilePopover from 'pnnl-react-core/lib/ProfilePopover';

const Home = ({ me }) => {
  if (me.id) {
    return (
      <Grid className="home-container">
        <h1>Home Page Title</h1>
        <p>You can load PNNL React Core components and even access the default user of your application. The boilerplate provides that base call for you.</p>
        <ProfilePopover person={me} />

        <ActionBar>
          <Col xs={4}>
            <Action to="/submit" icon="share" text="Submit" />
          </Col>
          <Col xs={4}>
            <Action to="/details" icon="info" text="Details" />
          </Col>
          <Col xs={4}>
            <Action to="/delete" icon="trash" text="Delete" />
          </Col>
        </ActionBar>
      </Grid>
    );
  }

  return (
    <div className="home-container">
      <h1>Hello!</h1>
      <p>No Data was fetched yet.</p>
    </div>
  );
};

Home.propTypes = {
  /**
    * A Person object for the person you want to create the card for.
    *
    *
    * The object needs to have the same naming convention as: https://api.pnl.gov/operations/v2/people/{hid}/?expand=space,facility
    * */
  me: PropTypes.shape({
    id: PropTypes.string,
    portraitLink: PropTypes.string,
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    title: PropTypes.string,
  }),
};

Home.defaultProps = {
  me: {},
};

const mapStateToProps = (state) => {
  const { me } = state;
  return {
    me,
  };
};

export default connect(mapStateToProps)(Home);
