// pnnl-react-core/lib/Tab/Tab.js
'use strict';

exports.__esModule = true;

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _NavItem = require('react-bootstrap/lib/NavItem');

var _NavItem2 = _interopRequireDefault(_NavItem);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _reactRouterBootstrap = require('react-router-bootstrap');

var _ValidationIcon = require('../ValidationIcon');

var _ValidationIcon2 = _interopRequireDefault(_ValidationIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Tab = function (_React$Component) {
  (0, _inherits3['default'])(Tab, _React$Component);

  function Tab() {
    (0, _classCallCheck3['default'])(this, Tab);
    return (0, _possibleConstructorReturn3['default'])(this, _React$Component.apply(this, arguments));
  }

  Tab.prototype.renderLink = function renderLink(tab) {
    var to = this.props.to;

    return _react2['default'].createElement(
      _reactRouterBootstrap.LinkContainer,
      { to: to },
      tab
    );
  };

  Tab.prototype.render = function render() {
    var _props = this.props,
        to = _props.to,
        onClick = _props.onClick,
        readonly = _props.readonly,
        children = _props.children,
        validationState = _props.validationState,
        props = (0, _objectWithoutProperties3['default'])(_props, ['to', 'onClick', 'readonly', 'children', 'validationState']);


    var tabClasses = (0, _classnames2['default'])('PnnlTabLabel', {
      validated: validationState !== undefined
    });

    var tab = _react2['default'].createElement(
      _NavItem2['default'],
      props,
      _react2['default'].createElement(
        'span',
        { className: tabClasses },
        children
      ),
      validationState && _react2['default'].createElement(_ValidationIcon2['default'], { validationState: validationState })
    );

    if (typeof to === 'string') {
      return this.renderLink(tab);
    }

    return tab;
  };

  return Tab;
}(_react2['default'].Component);

Tab.propTypes = {
  to: _propTypes2['default'].string,
  readonly: _propTypes2['default'].bool,
  onClick: _propTypes2['default'].func,
  validationState: _propTypes2['default'].oneOf(['success', 'error', 'warning', 'incomplete']),
  children: _propTypes2['default'].node.isRequired
};

Tab.defaultProps = {
  readonly: false,
  to: null,
  onClick: undefined,
  validationState: undefined
};

exports['default'] = Tab;
module.exports = exports['default'];
