// pnnl-react-core/lib/NavItem/NavItem.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Badge = require('react-bootstrap/lib/Badge');

var _Badge2 = _interopRequireDefault(_Badge);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Action = require('../Action');

var _Action2 = _interopRequireDefault(_Action);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var NavItem = function NavItem(_ref) {
  var count = _ref.count,
      rest = (0, _objectWithoutProperties3['default'])(_ref, ['count']);
  return _react2['default'].createElement(
    'li',
    { className: 'PnnlNavItem' },
    _react2['default'].createElement(_Action2['default'], (0, _extends3['default'])({ iconDisplay: 'all' }, rest)),
    count !== null && _react2['default'].createElement(
      _Badge2['default'],
      null,
      count
    )
  );
};

NavItem.propTypes = {
  /**
   * The name of the icon to be displayed in the NavItem.
   * @required
   */
  icon: _propTypes2['default'].string.isRequired,

  /**
   * The label to be displayed underneath the icon in the NavItem.
   * @required
   */
  text: _propTypes2['default'].string.isRequired,

  /**
   * The URL/route that the NavItem will link to.
   * @required
   */
  to: _propTypes2['default'].string.isRequired,

  /**
   * Determines whether the NavItem will display as 'active' when the 'to' path matches the loaded path exactly, or whether it matches any sub-path.  Example: when 'false', the NavItem will display as 'active' for both actions/ and action/submit. By default this is true. Talk to Geoff Elliott before changing this.
   * @private
   */
  onlyActiveOnIndex: _propTypes2['default'].bool,

  /**
   * A number to be displayed as a Badge on the NavItem. Should be used to notify or remind users of items they need to address. This should be a number, not a string or icon.
   */
  count: _propTypes2['default'].number
};

NavItem.defaultProps = {
  onlyActiveOnIndex: true,
  count: null
};

exports['default'] = NavItem;
module.exports = exports['default'];
