// pnnl-react-core/src/GoogleSearch/GoogleSearch.js
import React from 'react';
import { FormGroup, ControlLabel, FormControl } from 'react-bootstrap';
import classNames from 'classnames';
import Action from '../Action';
import Icon from '../Icon';

class GoogleSearch extends React.Component {
  constructor(props) {
    super(props);

    this.onTrigger = this.onTrigger.bind(this);

    this.state = {
      active: false, // Only applies to mobile, toggles the launch bar in/out
    };
  }

  onTrigger() {
    const newActiveState = !this.state.active;
    const searchInput = this.searchField;
    this.setState({
      active: newActiveState,
    });
    if (newActiveState) {
      searchInput.focus();
    } else {
      searchInput.blur();
    }
  }

  render() {
    const { active } = this.state;
    const searchBoxClasses = classNames(
      { active },
      'PnnlGoogleSearch',
    );

    return (
      <form className={searchBoxClasses} method="get" action="https://search.pnl.gov/search">
        <input type="hidden" name="as_sitesearch" value="" />
        <input type="hidden" name="access" value="a" />
        <input type="hidden" name="output" value="xml_no_dtd" />
        <input type="hidden" name="client" value="default_frontend" />
        <input type="hidden" name="site" value="default_collection" />
        <input type="hidden" name="proxystylesheet" value="default_frontend" />
        <input type="hidden" name="as_dt" value="i" />

        <Action icon="search" displayType="SearchTrigger" onClick={this.onTrigger} />
        <FormGroup>
          <ControlLabel title="Search" htmlFor="q"><Icon icon="search" /></ControlLabel>
          <FormControl type="text" name="q" id="q" ref={(e) => { this.searchField = e; }} placeholder="Search  Intranet" />
          <Action displayType="SearchClose" icon="crossmark" onClick={this.onTrigger} />
        </FormGroup>
      </form>
    );
  }
}

GoogleSearch.propTypes = {

};

export default GoogleSearch;
