// pnnl-react-core/src/Tab/Tab.js
import React from 'react';
import PropTypes from 'prop-types';
import NavItem from 'react-bootstrap/lib/NavItem';
import classNames from 'classnames';
import { LinkContainer } from 'react-router-bootstrap';
import ValidationIcon from '../ValidationIcon';


class Tab extends React.Component {
  renderLink(tab) {
    const { to } = this.props;
    return (
      <LinkContainer to={to}>
        {tab}
      </LinkContainer>
    );
  }

  render() {
    const {
      to, onClick, readonly, children, validationState, ...props
    } = this.props;

    const tabClasses = classNames(
      'PnnlTabLabel',
      {
        validated: (validationState !== undefined)
      }
    );

    const tab = (
      <NavItem {...props}>
        <span className={tabClasses}>
          {children}
        </span>
        {validationState && <ValidationIcon validationState={validationState} />}
      </NavItem>
    );

    if (typeof to === 'string') {
      return this.renderLink(tab);
    }

    return tab;
  }
}

Tab.propTypes = {
  to: PropTypes.string,
  readonly: PropTypes.bool,
  onClick: PropTypes.func,
  validationState: PropTypes.oneOf(['success', 'error', 'warning', 'incomplete']),
  children: PropTypes.node.isRequired,
};

Tab.defaultProps = {
  readonly: false,
  to: null,
  onClick: undefined,
  validationState: undefined,
};

export default Tab;
